vispy/visuals/border.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright (c) 2015, Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+# Author: Siddharth Bhat
+# -----------------------------------------------------------------------------
+
+import numpy as np
+
+from . import Visual
+from ..color import Color
+
+VERT_SHADER_BORDER = """
+attribute vec2 a_position;
+attribute vec2 a_adjust_dir;
+
+void main() {
+    // First map the vertex to document coordinates
+    vec4 doc_pos = $visual_to_doc(vec4(a_position, 0, 1));
+
+    // Also need to map the adjustment direction vector, but this is tricky!
+    // We need to adjust separately for each component of the vector:
+    vec4 adjusted;
+    if ( a_adjust_dir.x == 0 ) {
+        // If this is an outer vertex, no adjustment for line weight is needed.
+        // (In fact, trying to make the adjustment would result in no
+        // triangles being drawn, hence the if/else block)
+        adjusted = doc_pos;
+    }
+    else {
+        // Inner vertexes must be adjusted for line width, but this is
+        // surprisingly tricky given that the rectangle may have been scaled
+        // and rotated!
+        vec4 doc_x = $visual_to_doc(vec4(a_adjust_dir.x, 0, 0, 0)) -
+                    $visual_to_doc(vec4(0, 0, 0, 0));
+        vec4 doc_y = $visual_to_doc(vec4(0, a_adjust_dir.y, 0, 0)) -
+                    $visual_to_doc(vec4(0, 0, 0, 0));
+        doc_x = normalize(doc_x);
+        doc_y = normalize(doc_y);
+
+        // Now doc_x + doc_y points in the direction we need in order to
+        // correct the line weight of _both_ segments, but the magnitude of
+        // that correction is wrong. To correct it we first need to
+        // measure the width that would result from using doc_x + doc_y:
+        vec4 proj_y_x = dot(doc_x, doc_y) * doc_x;  // project y onto x
+        float cur_width = length(doc_y - proj_y_x);  // measure current weight
+
+        // And now we can adjust vertex position for line width:
+        adjusted = doc_pos + ($border_width / cur_width) * (doc_x + doc_y);
+    }
+
+    // Finally map the remainder of the way to render coordinates
+    gl_Position = $doc_to_render(adjusted);
+}
+"""
+
+FRAG_SHADER_BORDER = """
+void main() {
+    gl_FragColor = $border_color;
+}
+"""  # noqa
+
+
+class _BorderVisual(Visual):
+    """
+    Visual subclass to display 2D pixel-width borders.
+
+    Parameters
+    ----------
+    pos : tuple (x, y)
+        Position where the colorbar is to be placed with
+        respect to the center of the colorbar
+    halfdim : tuple (half_width, half_height)
```

Comment:
I thought I replied to @campagnola about that - that renaming it would
cause friction in the colorbar code (since it used the center
representation internally) with no real added benefit?

On Sat, Aug 15, 2015, 03:35 Eric Larson <notifications@github.com> wrote:

> In vispy/visuals/border.py
> <https://github.com/vispy/vispy/pull/1038#discussion_r37123883>:
>
> > +void main() {
> > +    gl_FragColor = $border_color;
> > +}
> > +"""  # noqa
> > +
> > +
> > +class _BorderVisual(Visual):
> > +    """
> > +    Visual subclass to display 2D pixel-width borders.
> > +
> > +    Parameters
> > +    ----------
> > +    pos : tuple (x, y)
> > +        Position where the colorbar is to be placed with
> > +        respect to the center of the colorbar
> > +    halfdim : tuple (half_width, half_height)
>
> @bollu <https://github.com/bollu> this hasn't been addressed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vispy/vispy/pull/1038/files#r37123883>.
>